File: priceserver/modules/sdr/fixer.py
import json
import requests
from typing import List, Dict

ENDPOINT = "http://data.fixer.io/api/latest"

def get_sdr_rates(currencies: List[str], api_key: str) -> Dict[str, float]:
    params = {
        # "base": "XDR",
        "access_key": api_key,
        "symbols": ','.join(currencies) + ',XDR'
    }

    result = requests.get(ENDPOINT, params=params).json()
    if not result['success']:
        raise NameError(result['error'])

    return change_base(result['rates'])

def change_base(rates: Dict[str, float]):
    sdr_rate = rates['XDR']
    del rates['XDR']

    for currency in rates:
        rates[currency] = sdr_rate / rates[currency]

    return rates

File: price-server/src/provider/fiat/Fixer.ts
import nodeFetch from 'node-fetch'
import * as logger from 'lib/logger'
import { errorHandler } from 'lib/error'
import { toQueryString } from 'lib/fetch'
import { num } from 'lib/num'
import { Quoter } from '../base'

interface Response {
  success: boolean
  rates: { [quote: string]: number }
  error?: string | Record<string, unknown>
}

export class Fixer extends Quoter {
  private async updatePrices(): Promise<void> {
    const params = {
      access_key: this.options.apiKey,
      // base: this.baseCurrency, // need 'PROFESSIONAL PLUS' subscription
      symbols: this.quotes.map((quote) => (quote === 'SDR' ? 'XDR' : quote)).join(',') + ',KRW',
    }

    const response: Response = await nodeFetch(
      `http://data.fixer.io/api/latest?${toQueryString(params)}`,
      {
        timeout: this.options.timeout,
      }
    ).then((res) => res.json())

    if (!response || !response.success || !response.rates) {
      logger.error(
        `${this.constructor.name}: wrong api response`,
        response ? JSON.stringify(response) : 'empty'
      )
      throw new Error('Invalid response from Fixer')
    }

    if (!response.rates.KRW) {
      throw new Error('there is no KRW price')
    }

    // convert to KRW prices
    const krwRate = 1 / +response.rates.KRW
    delete response.rates.KRW
    for (const quote of Object.keys(response.rates)) {
      response.rates[quote] *= krwRate
    }

    // update last trades
    for (const quote of Object.keys(response.rates)) {
      this.setPrice(quote === 'XDR' ? 'SDR' : quote, num(response.rates[quote]))
    }
  }

  protected async update(): Promise<boolean> {
    await this.updatePrices().catch(errorHandler)

    return true
  }
}

export default Fixer
